File: app/android/app/src/main/java/com/barghest/bux/data/dto/Budget.kt
package com.barghest.bux.data.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class BudgetResponse(
    val id: Int,
    @SerialName("category_id") val categoryId: Int,
    val category: CategoryResponse? = null,
    val amount: String,
    val currency: String,
    val period: String
)

@Serializable
data class CreateBudgetRequest(
    @SerialName("category_id") val categoryId: Int,
    val amount: String,
    val currency: String = "RUB",
    val period: String = "monthly"
)

@Serializable
data class UpdateBudgetRequest(
    val amount: String? = null,
    val period: String? = null
)

@Serializable
data class BudgetStatusResponse(
    @SerialName("budget_id") val budgetId: Int,
    @SerialName("category_id") val categoryId: Int,
    @SerialName("category_name") val categoryName: String,
    @SerialName("category_icon") val categoryIcon: String,
    @SerialName("category_color") val categoryColor: String,
    @SerialName("budget_amount") val budgetAmount: String,
    @SerialName("spent_amount") val spentAmount: String,
    val remaining: String,
    @SerialName("spent_percent") val spentPercent: String,
    val period: String,
    val currency: String
)

File: server/services/transaction/internal/presentation/http/dto/budget.go
package dto

import "transaction/internal/domain/model"

type BudgetResponse struct {
	ID         uint              `json:"id"`
	CategoryID uint              `json:"category_id"`
	Category   *CategoryResponse `json:"category,omitempty"`
	Amount     string            `json:"amount"`
	Currency   string            `json:"currency"`
	Period     string            `json:"period"`
}

type CreateBudgetRequest struct {
	CategoryID uint   `json:"category_id" binding:"required"`
	Amount     string `json:"amount" binding:"required"`
	Currency   string `json:"currency"`
	Period     string `json:"period"`
}

type UpdateBudgetRequest struct {
	Amount string `json:"amount"`
	Period string `json:"period"`
}

type BudgetStatusResponse struct {
	BudgetID      uint   `json:"budget_id"`
	CategoryID    uint   `json:"category_id"`
	CategoryName  string `json:"category_name"`
	CategoryIcon  string `json:"category_icon"`
	CategoryColor string `json:"category_color"`
	BudgetAmount  string `json:"budget_amount"`
	SpentAmount   string `json:"spent_amount"`
	Remaining     string `json:"remaining"`
	SpentPercent  string `json:"spent_percent"`
	Period        string `json:"period"`
	Currency      string `json:"currency"`
}

func BudgetFromModel(b model.Budget) BudgetResponse {
	resp := BudgetResponse{
		ID:         b.ID,
		CategoryID: b.CategoryID,
		Amount:     b.Amount.String(),
		Currency:   b.Currency,
		Period:     string(b.Period),
	}
	if b.Category != nil {
		cat := CategoryFromModel(*b.Category)
		resp.Category = &cat
	}
	return resp
}

func BudgetListFromModel(budgets []model.Budget) []BudgetResponse {
	res := make([]BudgetResponse, len(budgets))
	for i, b := range budgets {
		res[i] = BudgetFromModel(b)
	}
	return res
}
